lib/gitgitgadget.ts

```diff
@@ -205,16 +206,25 @@ export class GitGitGadget {
             this.publishTagsAndNotesToRemote,
             "--",
             `+${this.notes.notesRef}:${this.notes.notesRef}`,
-            `+${pullRequestRef}:${pullRequestRef}`,
-            `+${pullRequestMerge}:${pullRequestMerge}`,
             `+refs/heads/maint:refs/remotes/upstream/maint`,
             `+refs/heads/master:refs/remotes/upstream/master`,
             `+refs/heads/next:refs/remotes/upstream/next`,
             `+refs/heads/pu:refs/remotes/upstream/pu`,
         ];
+        const prArgs = [
+            `+${pullRequestRef}:${pullRequestRef}`,
+            `+${pullRequestMerge}:${pullRequestMerge}`,
+        ];
         if (additionalRef) {
             args.push(`+${additionalRef}:${additionalRef}`);
         }
+        if (repositoryOwner === "gitgitgadget") {
+            args.push(...prArgs);
+        } else {
+            prArgs.unshift("fetch", `https://github.com/${repositoryOwner}/git`,
+                           "--");
+            await git(prArgs, { workDir: this.workDir });
```

Comment:
Sorry if this is a silly question.  How tightly in synch are gitgitgadget/git and git/git?  Just want to make sure `pr.baseCommit` will always be in the gitgitgadget repo for someone opening a PR on the git repo.

Reply:
> Sorry if this is a silly question. How tightly in synch are gitgitgadget/git and git/git?

I am glad you asked: gitgitgadget/git is kept up to date by two Azure Pipelines: [Synchronize gitster.git to GitGitGadget](https://dev.azure.com/gitgitgadget/git/_build?definitionId=8) and [Synchronize git.git to GitGitGadget](https://dev.azure.com/gitgitgadget/git/_build?definitionId=7).

The existence of the `git/git` -> `gitgitgadget/git` pipeline is probably obvious.

The `gitster/git` repository contains the individual branches for code contributions, and the first reason why I added the corresponding pipeline was that I got a bit sick and tired of `pu` failing all the time without any actionable information: it has been always unclear _which_ of those patch series that made it into `pu` was responsible for the test failures. Now with the individual branches being mirrored into `gitgitgadget/git`, and obviously triggering [CI builds](https://dev.azure.com/gitgitgadget/git/_build?definitionId=4), it is a lot easier to find the culprits.

Of course, from time to time the `pu` failures are caused by unfortunate interactions of separate patch series' changes, in which case the CI builds of the individual branches may very well succeed but `pu` still causes failures, which is a very useful piece of information in and of itself.

A secondary benefit of mirroring the `gitster/git` branches is that PRs at `gitgitgadget/git` can target more fine-grained base commits. I use this a lot, e.g. in my six "built-in add -i/-p" patch series which build on top of each other.

The pipeline that mirrors gitster/git into gitgitgadget/git (the pipeline names cannot contain slashes, that's why the `/` was replaced by a `.`) has two tasks:

1. Synchronize branch (the pipeline is triggered by all updates to `refs/heads/*`, polling every 180 seconds, i.e. every three minutes):
   ```bash
   case "$(Build.SourceBranch)" in
   refs/heads/*)
       refspec="+HEAD:$(Build.SourceBranch)"
       ;;
   refs/tags/*)
       refspec="$(Build.SourceBranch)"
       ;;
   *)
       echo "Cannot handle '$(Build.SourceBranch)'" >&2
       exit 1
       ;;
   esac

   git -c  http."https://github.com/gitgitgadget/git".extraheader="Authorization: Basic $(gitgitgadget.push.token.base64)" \
       push https://github.com/gitgitgadget/git "$refspec"
   ```
2. Synchronize tags (if necessary):
   ```bash
   die () {
       echo "$*" >&2
       exit 1
   }

   for d in git gitgitgadget
   do
       git ls-remote --tags https://github.com/$d/git | grep -v '\^{}$' | sort >tags.$d ||
       die "Could not enumerate tags in $d"
   done

   refspec="$(comm -23 tags.git tags.gitgitgadget | tr '\t' :)" ||
   die "Could figure out missing tags"

   if test -z "$refspec"
   then
       echo "##vso[task.complete result=Skipped]No tags to synchronize!"
   else
       git -c  http."https://github.com/gitgitgadget/git".extraheader="Authorization: Basic $(gitgitgadget.push.token.base64)" \
           push https://github.com/gitgitgadget/git $refspec
   fi
   ```

That second task is necessary because there is currently no way to trigger non-YAML Azure Pipelines from tag updates. Of course this means that new tags are only synchronized together with branch updates, but in practice that's okay because Junio always pushes out the tags with corresponding branches.

You will also notice that this task mirrors tags from the git/git repository, even if the pipeline purports to mirror gitster/git to gitgitgadget/git; This is a deliberate decision, to make sure that we only mirror the "official tags" from the authoritative repisotyr.

The pipeline that synchronizes git/git to gitgitgadget/git predates the other pipeline, and is probably made somewhat redundant by it. I still keep it around, and not only for sentimental reasons: it used to poll every 1800 seconds (i.e. every half hour), but the gitgitgadget builds recently got access to the webhooks, so it no longer polls (I *just* changed it, having noticed the behavior while writing this answer, and while at it, I also added a task that synchronizes the tags, which had been previously missing from that pipeline).

Side note: there is also the [Synchronize git-gui.git (branches only) to GitGitGadget](https://dev.azure.com/gitgitgadget/git/_build?definitionId=10) pipeline. Its purpose is obvious: to mirror the branches of Pratyush Yadav's repository (i.e. the current Git GUI maintainer's authoritative repository) into the `git-gui/*` namespace on gitgitgadget/git.

Reply:
I copy-edited this into https://github.com/gitgitgadget/gitgitgadget/wiki/GitGitGadget's-Azure-Function-and-Azure-Pipelines.

Reply:
Thanks for the detailed explanation and adding it to the wiki.  Always good to have the info in case you take vacation. 😀